sessionctx/variable/session.go

```diff
@@ -1426,34 +1429,167 @@ type TableDelta struct {
 	InitTime time.Time // InitTime is the time that this delta is generated.
 }
 
+// NullString means the value is unset.
+const NullString = ""
+
+// DeprecatedConcurrencyVariables hold the name with default value for deprecated concurrency variables.
+var DeprecatedConcurrencyVariables = map[string]string{
+	TiDBIndexLookupConcurrency:     strconv.Itoa(DefIndexLookupConcurrency),
+	TiDBIndexLookupJoinConcurrency: strconv.Itoa(DefIndexLookupJoinConcurrency),
+	TiDBHashJoinConcurrency:        strconv.Itoa(DefTiDBHashJoinConcurrency),
+	TiDBHashAggFinalConcurrency:    strconv.Itoa(DefTiDBHashAggFinalConcurrency),
+	TiDBHashAggPartialConcurrency:  strconv.Itoa(DefTiDBHashAggPartialConcurrency),
+	TiDBWindowConcurrency:          strconv.Itoa(DefTiDBWindowConcurrency),
+	TiDBProjectionConcurrency:      strconv.Itoa(DefTiDBProjectionConcurrency),
+	TiDBDistSQLScanConcurrency:     strconv.Itoa(DefDistSQLScanConcurrency),
+}
+
 // Concurrency defines concurrency values.
 type Concurrency struct {
-	// IndexLookupConcurrency is the number of concurrent index lookup worker.
-	IndexLookupConcurrency int
+	// indexLookupConcurrency is the number of concurrent index lookup worker.
+	indexLookupConcurrency *int
+
+	// indexLookupJoinConcurrency is the number of concurrent index lookup join inner worker.
+	indexLookupJoinConcurrency *int
+
+	// distSQLScanConcurrency is the number of concurrent dist SQL scan worker.
+	distSQLScanConcurrency *int
+
+	// hashJoinConcurrency is the number of concurrent hash join outer worker.
+	hashJoinConcurrency *int
+
+	// projectionConcurrency is the number of concurrent projection worker.
+	projectionConcurrency *int64
+
+	// hashAggPartialConcurrency is the number of concurrent hash aggregation partial worker.
+	hashAggPartialConcurrency *int
+
+	// hashAggFinalConcurrency is the number of concurrent hash aggregation final worker.
+	hashAggFinalConcurrency *int
+
+	// windowConcurrency is the number of concurrent window worker.
+	windowConcurrency *int
+
+	// indexSerialScanConcurrency is the number of concurrent index serial scan worker.
+	indexSerialScanConcurrency int
```

Comment:
These should be documented as deprecated.

Reply:
done